lib/auth/auth_with_roles.go

```diff
@@ -3531,7 +3531,17 @@ func (a *ServerWithRoles) GetOIDCAuthRequest(ctx context.Context, id string) (*t
 
 func (a *ServerWithRoles) ValidateOIDCAuthCallback(ctx context.Context, q url.Values) (*OIDCAuthResponse, error) {
 	// auth callback is it's own authz, no need to check extra permissions
-	return a.authServer.ValidateOIDCAuthCallback(ctx, q)
+	resp, err := a.authServer.ValidateOIDCAuthCallback(ctx, q)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Only the Proxy service can create web sessions via OIDC connector.
+	if resp.Session != nil && !a.hasBuiltinRole(types.RoleProxy) {
+		return nil, trace.AccessDenied("this request can be only executed by a proxy")
```

Comment:
Should we also delete the session? Something like `a.WebSession().Delete(ctx, ...)`

Reply:
Good point. I moved the check to `CreateXXXAuthRequest` so we can check before creating the web session rather than deleting it.

Reply:
Actually, I've just put the check in both places for redundancy. The session will automatically expire on its own, and this path should be practically unreachable, so I think it's ok not to delete it explicitly. Ideally we should check before creating the web session but it's a non trivial refactor.